tssc/step_implementers/uat/maven.py

```diff
@@ -0,0 +1,247 @@
+        attempt to pull it from a previous deploy step.
+
+        If still unable to find, then throw ValueError.
+        """
+        target_base_url = self.get_config_value('target-base-url')
+        if target_base_url is None:
+            deploy_config = self.get_step_results(DefaultSteps.DEPLOY)
+            if deploy_config is not None:
+                target_base_url = deploy_config.get('deploy-endpoint-url')
+
+        if target_base_url is None:
+            raise ValueError('No target base url was specified')
+
+        return target_base_url
+
+    def _run_step(self):
+        """
+        Runs the TSSC step implemented by this StepImplementer.
+
+        Parameters
+        ----------
+        runtime_step_config : dict
+            Step configuration to use when the StepImplementer runs the step with all of the
+            various static, runtime, defaults, and environment configuration munged together.
+
+        Returns
+        -------
+        dict
+            Results of running this step.
+        """
+        pom_file = self.get_config_value('pom-file')
+        fail_on_no_tests = self.get_config_value('fail-on-no-tests')
+        selenium_hub_url = self.get_config_value('selenium-hub-url')
+        report_dir = self.get_config_value('report-dir')
+        target_base_url = self._get_target_base_url()
+
+        if not os.path.exists(pom_file):
+            raise ValueError(f'Given pom file does not exist: {pom_file}')
+
+        surefire_path = 'mvn:build/mvn:plugins/mvn:plugin/[mvn:artifactId="maven-surefire-plugin"]'
+        maven_surefire_plugin = get_xml_element_by_path(
+            pom_file,
+            surefire_path,
+            default_namespace='mvn'
+        )
+        if maven_surefire_plugin is None:
+            raise ValueError('Uat dependency "maven-surefire-plugin" missing from POM.')
+
+        reports_dir = get_xml_element_by_path(
+            pom_file,
+            f'{surefire_path}/mvn:configuration/mvn:reportsDirectory',
+            default_namespace='mvn'
+        )
+        if reports_dir is not None:
+            test_results_dir = reports_dir.text
+        else:
+            test_results_dir = os.path.join(
+                os.path.dirname(os.path.abspath(pom_file)),
+                'target/surefire-reports')
+
+        try:
+            sh.mvn(  # pylint: disable=no-member
+                'clean',
+                '-Pintegration-test',
+                f'-Dselenium.hub.url={selenium_hub_url}',
+                f'-Dtarget.base.url={target_base_url}',
+                f'-Dcucumber.plugin=html:target/{report_dir}/cucumber.html,' \
+                    f'json:target/{report_dir}/cucumber.json',
+                'test',
+                '-f', pom_file,
+                _out=sys.stdout
+            )
+        except sh.ErrorReturnCode as error:
+            raise RuntimeError("Error invoking mvn: {error}".format(error=error)) from error
```

Comment:
IF you make changes to this file, possibly update to the new string format pattern?
raise RuntimeError(f"Error invoking mvn: {error}") from error

Reply:
Ah ya good call.  I prefer the `f` string over `format` too.

Reply:
+1